### PR TITLE
sql: add information_schema._pg_expandarray

### PR DESCRIPTION
**Pull request by mjibson**
Add the information_schema schema, as this is its first function.

Add a new Func variant that's similar to table funcs, but directly
exposes a relation expr and scope. Add a function that can produce the
appropriate expr given a SQL statement, allowing us to implement this
function in a way similar to Postgres.

Although we have added support for _pg_expandarray, it's still not yet
enough to support PgJDBC, which needs #1546. However we can add the
FROM-side support for this function now, and separately implement 1546
which will turn this on.

### Motivation

  * This PR adds a known-desirable feature. #4506

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR adds a release note for any
  [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

**Review by sploiselle — APPROVED** [review]
LGTM modulo comment about `sql_impl_table`, whose distinction from `sql_impl_table_func` isn't totally clear to me.

File: src/sql/src/func.rs
Comment (mjibson) [reply src/sql/src/func.rs:335]: I was copying the other sql funcs, but those are used in two places. I've merged these into just one.

File: src/sql/src/func.rs
Comment (mjibson) [reply src/sql/src/func.rs:367]: done

File: src/sql/src/func.rs
Comment (mjibson) [reply src/sql/src/func.rs:2044]: I think you are correct. It is this way because that's how it is in postgres (https://github.com/postgres/postgres/blob/16e3ad5d143795b05a21dc887c2ab384cce4bcb8/src/backend/catalog/information_schema.sql#L43) and I suppose it'll also be future-proof if we ever support the weirder array syntax.

Comment (sploiselle) [reply src/sql/src/func.rs]: Cool! Just wanted to make sure my thinking was clear.

**Comment by benesch** [comment]
I defer to Sean!

**Comment by mjibson** [comment]
@philip-stoev Can you look into the upgrade test failure? It's a real failure and happens locally too, but I have no idea what it means. I've tried to debug it locally but am unable to make progress. I have by hand created a 0.8.0 mzdata directory from our release binary, then ran this on top of it, and it successfully applies the migration, so I'm pretty confused. The error message referencing the create-in files also baffles me, since those should run before any new code.

**Comment by benesch** [comment]
Could be what @cjubb39 just fixed in
https://github.com/MaterializeInc/materialize/commit/488d4d62639287aa7536721cc3cb7fcbcd78686a
!

**Comment by philip-stoev** [comment]
Yes indeed this is the same issue @cjubb39  fixed in his branch. If you need to push this PR right away, the following quick-and-dirty patch will get the upgrade test to pass:

```
diff --git a/misc/python/materialize/mzcompose.py b/misc/python/materialize/mzcompose.py
index 939ec6184..5b99975b0 100644
--- a/misc/python/materialize/mzcompose.py
+++ b/misc/python/materialize/mzcompose.py
@@ -914,7 +914,6 @@ class Testdrive(PythonService):
             "--kafka-addr=kafka:9092",
             "--schema-registry-url=http://schema-registry:8081",
             "--materialized-url=postgres://materialize@materialized:6875",
-            "--validate-catalog=/share/mzdata/catalog",
             "${TD_TEST:-$$*}",
         ],
         environment: List[str] = [
```

(@cjubb39 will then get a hopefully auto-resolveable merge conflict on his side)

**Comment by cjubb39** [comment]
so coincidental that we ran into this at the same time.  My PR has merged so, if it was the same root issue, should be fixed by a rebase!

**Pull request merged**
